### PR TITLE
Use ANTLR for validation rules DHIS2-3628

### PR DESCRIPTION
**Pull request by jimgrace**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java [hunk 1]
```diff
@@ -250,41 +258,58 @@ public void setUpTest()
         programStageInstanceService.addProgramStageInstance( stageInstanceA );
         programStageInstanceService.addProgramStageInstance( stageInstanceB );
 
+//        IndicatorType indicatorTypeA = createIndicatorType( 'A' );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Remove commented out code.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java [hunk 2]
```diff
@@ -250,41 +258,58 @@ public void setUpTest()
...
 
         expressionService.addExpression( expressionA );
         expressionService.addExpression( expressionD );
         expressionService.addExpression( expressionI );
+//        expressionService.addExpression( expressionN );
 
         validationRuleA = createValidationRule( "A", not_equal_to, expressionA, expressionA, periodTypeMonthly ); // A - ProgramTrackedEntityAttribute
         validationRuleD = createValidationRule( "D", not_equal_to, expressionD, expressionD, periodTypeMonthly ); // D - ProgramDataElement
         validationRuleI = createValidationRule( "I", not_equal_to, expressionI, expressionI, periodTypeMonthly ); // I - ProgramIndicator
+//        validationRuleN = createValidationRule( "N", not_equal_to, expressionN, expressionN, periodTypeMonthly ); // N - Indicator
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Remove commented out code.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java [hunk 3]
```diff
@@ -342,6 +368,12 @@ private Grid newGrid( double dataElementVal, double teAttributeVal, double piVal
         grid.addValue( "HllvX50cXC0" );
         grid.addValue( new Double( piVal ) );
 
+//        grid.addRow();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Remove commented out code.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java [hunk 4]
```diff
@@ -407,10 +441,10 @@ private void assertResultsEquals( Collection<ValidationResult> reference, Collec
             sb.append( formatResultsList( "Expected", referenceList ) )
                 .append( formatResultsList ( "But was", resultsList ) );
 
-            log.error( sb.toString() );
+            System.err.println( sb.toString() );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Remove system.err.println.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java [hunk 5]
```diff
@@ -443,12 +477,16 @@ public void testEventValidate()
         reference.add( new ValidationResult( validationRuleD, periodApr, orgUnitA, defaultCombo, 5.0, 5.0, dayInPeriod ) );
         reference.add( new ValidationResult( validationRuleI, periodMar, orgUnitA, defaultCombo, 8.0, 8.0, dayInPeriod ) );
         reference.add( new ValidationResult( validationRuleI, periodApr, orgUnitA, defaultCombo, 9.0, 9.0, dayInPeriod ) );
+//        reference.add( new ValidationResult( validationRuleN, periodMar, orgUnitA, defaultCombo, 3.0, 3.0, dayInPeriod ) );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Remove commented out code.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java (hunk 1 above)
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Fixed.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java (hunk 2 above)
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Fixed.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java (hunk 3 above)
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Fixed.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java (hunk 4 above)
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Fixed.

File: dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java (hunk 5 above)
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/test/java/org/hisp/dhis/validation/AnalyticsValidationServiceTest.java]: Fixed.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/validation/PeriodTypeExtended.java [hunk 6]
```diff
@@ -128,6 +140,18 @@ public String getDeoIds( long dataElementId, long categoryOptionComboId )
         return dataElementId + "." + categoryOptionComboId;
     }
 
+    public void setSlidingWindows( boolean slidingWindows )
+    {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/validation/PeriodTypeExtended.java]: This part is a bit excessive. If "slidingWindowsNeeded"/"nonSlidingWindowsNeeded" are determined by "slidingWindows" only, we could have just one stateful boolean property `slidingWindows`, with getters for:

```
public boolean slidingWindowsNeeded() 
{
  return slidingWindows;
}

public boolean nonSlidingWindowsNeeded()
{
  return !slidingWindows;
}
```

I.e. better to have as few stateful properties as possible / less properties to keep in sync.

Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/validation/PeriodTypeExtended.java]: When initializing a PeriodTypeExtended, are three possibilities: (1) all validation rule expressions in the periodType have slidingWindows = true, resulting in slidingWindowsNeeded = true and nonSlidingWindowsNeeded = false, (2) all validation rule expressions in that periodType have slidingWindows = false, resulting in slidingWindowsNeeded = false and nonSlidingWindowsNeeded = true, or (3) there are some of each, resulting in slidingWindowsNeeded = true and nonSlidingWindowsNeeded = true.

This is a performance improvement. The old code always called analytics to get the sliding windows data, and also always called analytics to get the non-sliding windows data. The new code only calls analytics for the sliding windows if needed, and only calls analytics for the non-sliding windows if needed.

Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/validation/PeriodTypeExtended.java]: Okay makes sense.

**Pull request merged**